kazoo/protocol/connection.py

```diff
@@ -686,68 +674,112 @@ def _connect(self, host, port):
-                    if scheme == 'sasl':
-                        username, password = auth.split(":")
-                        self.sasl_cli = SASLClient(
-                            host=client.sasl_server_principal,
-                            service='zookeeper',
-                            mechanism='DIGEST-MD5',
-                            username=username,
-                            password=password
-                        )
-                        break
-
-                # As described in rfc
-                # https://tools.ietf.org/html/rfc2831#section-2.1
-                # sending empty challenge
-                self._send_sasl_request(challenge=b'',
-                                        timeout=connect_timeout)
-            else:
-                self.logger.warn('Pure-sasl library is missing while sasl'
-                                 ' authentification is configured. Please'
-                                 ' install pure-sasl library to connect '
-                                 'using sasl. Now falling back '
-                                 'connecting WITHOUT any '
-                                 'authentification.')
-                client.use_sasl = False
-                self._set_connected_ro_or_rw(client)
-        else:
-            self._set_connected_ro_or_rw(client)
-            for scheme, auth in client_auth_data_copy:
-                if scheme == "digest":
-                    ap = Auth(0, scheme, auth)
-                    zxid = self._invoke(
-                        connect_timeout / 1000.0,
-                        ap,
-                        xid=AUTH_XID
-                    )
-                    if zxid:
-                        client.last_zxid = zxid
+        for scheme, auth in client_auth_data_copy:
+            ap = Auth(0, scheme, auth)
+            zxid = self._invoke(connect_timeout / 1000.0, ap, xid=AUTH_XID)
+            if zxid:
+                client.last_zxid = zxid
 
         return read_timeout, connect_timeout
 
-    def _send_sasl_request(self, challenge, timeout):
-        """ Called when sending a SASL request, xid needs be to incremented """
-        sasl_request = SASL(challenge)
-        self._xid = (self._xid % 2147483647) + 1
-        xid = self._xid
-        self._submit(sasl_request, timeout / 1000.0, xid)
-
-    def _set_connected_ro_or_rw(self, client):
-        """ Called to decide whether to set the KeeperState to CONNECTED_RO
-            or CONNECTED"""
-        if self._ro:
-            client._session_callback(KeeperState.CONNECTED_RO)
-            self._ro_mode = iter(self._server_pinger())
-        else:
-            client._session_callback(KeeperState.CONNECTED)
-            self._ro_mode = None
+    def _authenticate_with_sasl(self, host, timeout):
+        """Establish a SASL authenticated connection to the server.
+        """
+        if not PURESASL_AVAILABLE:
+            raise SASLException('Missing SASL support')
```

Comment:
As it is currently done, isn't it better to connect to the server without any auth and inform the client with a warning message?

Reply:
My line of thought is that if I create a ZK client object with auth data and that auth data is rejected by the server, it feels to me that I provided "wrong data" and it should tell me right away.

Another way to look at this is that this is an "auth error" while trying to access a node you do not have permission to (which would likely happen if you failed to authenticate) is a "perm error".  Making the authentication failure silent would, in effect, alias the two kind of errors and make the ZK client less expressive.

Lastly, it would be trivial for an application to simply catch the auth error, reconfigure the ZK client *without* the bad auth data and reconnect, to emulate the silent continue.
On the other hand, it would not be as easy to connect and then check if we actually successfully authenticated.

That being said, in my work environment, authentication is pretty much mandatory, so my point of view might be skewed here.

Reply:
You are right, it makes sense, let's keep it the way you did.